### ISSUE TITLE
Quick distance measurement using two fingers taps

### ISSUE THREAD

**Issue by pastk** [organicmaps/organicmaps#1867]
Some other mapping apps (e.g. mapy.cz, 2gis) have this feature and its great.

Just tap two different points on the map at the same time and the app will show the straight line distance between them. It disappears as soon as fingertips are released.

![2tap-distance](https://user-images.githubusercontent.com/18434508/149654360-0e9d9e2e-1260-4475-8539-486f8e0dc11b.png)

I think this feature is very much in the vein of OM approach - simple, efficient, doesn't get into one's way, minimum clicks required. Its literally like having distances at one's fingertips :)
And people who don't need it won't even notice its existence (no visible UI change for them).

I think its not duplicate, but complimentary to #1452.
Two-fingers approach would cover 90% of measurement needs I think.
More rare cases like measurement between several points, high measurement precision needs and straight line navigation could be covered by the "helicopter routing" at the expense of many more clicks and more intrusive experience overall.

**Comment by biodranik** [comment]
The problem is how to make it discoverable. We can't teach users now about a long tap :)

**Comment by pastk** [comment]
> The problem is how to make it discoverable. We can't teach users now about a long tap :)

True.
Though this particular feature will be visible for users who use 2-fingers zoom and rotate gestures (well, depending on implementation).

**Comment by pastk** [comment]
mapy.cz implementation:
![mapy_cz](https://user-images.githubusercontent.com/18434508/149757038-ea278bd6-bda5-46e3-bd9c-511bb142c3b1.jpg)

https://user-images.githubusercontent.com/18434508/149757087-f2261dcf-5377-4d31-812a-8ffec3e9164b.mp4

2gis implementation:




https://user-images.githubusercontent.com/18434508/149757155-2c7c5d1e-1ee8-43ac-8a7e-1d8a6c2f49dc.mp4

I think mapy.cz-like simpler implementation is a better fit for OM.

**Comment by biodranik** [comment]
Helicopter/crow routing will be discoverable and easy to use, with the exact same interface as for any other type of routing, with a bonus of the possibility to measure more than one segment distance.

**Comment by vanous** [comment]
Thank you for all the hard work on Organic maps. I would love to see this feature but I am not here to advocate for it being implemented, but to give reasoning why (if implemented) it is so awesome to have. mapy.cz have this and I had no idea for many years, so definitely hard to discover. Not sure how to solve this.

But, once you know about it you just miss it badly - it doesn't require setting two points, you casually look at the map, place two fingers and bingo - right away you get a rough idea about how a city or forest is long. Or, how far something is.

Anyways, thank you for this awesome work and don't get burned out.

**Comment by biodranik** [comment]
I agree, it's convenient if it's done in unobtrusive way. A bit hard to discover though.

**Comment by vanous** [comment]
Actually, I re-tried normal routing (with the helicopter routing on mind) and if the helicopter route is added, it can actually be a good replacement, as the from/to setup flow is quite fast and straight forward. It takes a few more taps but not that much and it solves the discovery issue.

**Comment by alatarum** [comment]
> Just tap two different points on the map at the same time and the app will show the straight line distance between them. 

> And people who don't need it won't even notice its existence (no visible UI change for them).

These two statements sounds contradictory to me: first one means that when I use two fingers to zoom/rotate/scroll,distance will immediately appears and overlay map information - this is not something that I expect 90% of time.

The Guru Maps implementation looks good and, BTW, could be nice integrated with helicopter routing, like this:
1) do long double tap - distance appear together with helicopter routing button
2) pressing the button opens helicopter routing interface with start and end points already selected
3) now it is possible to move points with tap&hold and add new points.

**Comment by pastk** [comment]
> These two statements sounds contradictory to me: first one means that when I use two fingers to zoom/rotate/scroll,distance will immediately appears and overlay map information - this is not something that I expect 90% of time.

Check the mapy.cz and Guru implementations, a user need to wait a fraction of a second for the line to appear.

However I prefer an idea of immediate ruler appearance - it solves the issue with discoverability of the feature.

The Guru implementation has a nice benefit of precision, but if the ruler will appear immediately some users will get confused whats there staying on the screen after they finished zooming/rotating.

**Comment by alatarum** [comment]
Well, I really don't like idea of immediate ruler appearance: it would be unwanted behavior with overlapping map data with distance most of the time. I trust, users can discover long tap actions if they need something like this =)

**Comment by pastk** [comment]
> I trust, users can discover long tap actions if they need something like this =)

No, unfortunately https://github.com/organicmaps/organicmaps/issues/1867#issuecomment-1013972075
Discoverability was on of the main reasons to select arbitrary map points via short tap instead of a long tap.

**Comment by JStrm** [comment]
+1 I would love to see this feature

**Comment by vanous** [comment]
It has already been implemented and it's actually much better than on mapy.cz

Tap on the map choose route from, tap again on the map choose route to. Select the "direct line" (as the crow flying).

You can now keep adding removing and reordering the points. 

I think this issue can be closed now.

**Comment by Misalf-git** [comment]
> I think this issue can be closed now.

This is not about just any way of measuring, but specifically "**Quick distance measurement using two fingers**" (without multiple button taps and/or cancelling running actions like navigation).

**Comment by vanous** [comment]
I had used that feature on maps.cz many , many times and replicating it here would actually be downgrade from what we already have. On mapy, many times it doesn't work because you need to hit the screen with both fingers in exactly the same time. And once you do, often the location is not exactly where you really wanted it to be. But, that's my experience, feel free to advocate for your use case.

**Comment by hoizan** [comment]
I prefer the current route-to method. Not only it does distance measurements elegantly, it's also great for route planning and  good for navigation too if one prefer unguided direct free navigating especially for orienteering events.

**Comment by JStrm** [comment]
> It has already been implemented and it's actually much better than on mapy.cz

I have to disagree, I use the feature to quickly gauge how far two points are, and the implementation in Maps.com is much better at this. When walking, i use it to very quickly determine how far away i am from some point, or how far away some other 2 points are. If I have the time to plan a route, i wouldn't bother choosing the straight line path, as the walking/biking/driving distance is much more useful at that point. 

I personally don't see the point of measuring the distance using the straight line when other, more accurate options are available. I trust that for other users, this feature may be useful, but not for me.

**Comment by biodranik** [comment]
>  use it to very quickly determine how far away i am from some point

Select any point on the map to immediately see the distance from you to it.

**Comment by potatopplking** [comment]
Hello! I'm a long time user of mapy.cz, just started to use Organic Maps recently. Great app, but I'm really missing that two-finger distance measurement - it is faster and more convenient. I did quick proof-of-concept implementation and it works for me (long-tap with two fingers show the distance, otherwise the behavior of the UI is unchanged), but I'm not sure if it's worth trying to make PR here. Now the code is definitely not presentation-worthy :) and I can see there's already plenty of open PRs here, so I don't want to bother you with another one.
 
Anyway, here's the screenshot, if there would be a chance of this feature getting accepted, I will gladly put more effort into it.

<img width="540" height="1200" alt="Image" src="https://github.com/user-attachments/assets/13896a45-9e82-4d92-8974-d80a59307976" />

Thanks for the great app and all the work you put into it.

**Comment by biodranik** [comment]
@potatopplking thanks for the POC, the main issue with mapy is the gesture itself: it is unstable, and doesn't work reliably sometimes. If it works great in OM, make a PR and we can beta-test it on iOS and Android.
